async_service/asyncio.py

```diff
@@ -272,43 +273,77 @@ def run_child_service(
         return child_manager
 
 
-TReturn = TypeVar("TReturn")
-TFunc = TypeVar("TFunc", bound=Callable[..., Coroutine[Any, Any, TReturn]])
+def cleanup_tasks(*tasks: "asyncio.Future[Any]") -> AsyncContextManager[None]:
```

Comment:
Is this wrapper around `_cleanup_tasks()` just dealing with some mypy shenanigans?

Reply:
AIUI (and it took me a really long time before I did), this is a trick to "consume" the first task and have `_cleanup_tasks()` called with only the remaining ones. Otherwise, on line 294 we'd end up always calling it with the same sequence of tasks. It is a clever trick but doesn't come naturally to me (and I guess to you either?), so I can't help but wonder if it wouldn't be better to write it in a less clever but more obvious way?

Reply:
I think it was actually both of these things.  Previously, I needed a `type: ignore` on the `_cleanup_tasks` implementation (for some reason that is no longer the case either because `mypy` got smarter or something else).  That went away at some point but, @gsalgado is also correct that this was a way to consume the first element of the list of tasks without needing to do any length checking or error handling because it allows it to be handled at the language level.

I'll look at how to clean this up, either by doing it less cleverly or potentially by adding some comments to explain it clearly.